### PR TITLE
Update testing framework to modern Chef standards

### PR DESCRIPTION
**Pull request by wolf31o2**
This is a massive update to bring this cookbook up to modern Chef
standards. It requires ChefDK to be installed and assumes its use,
which is the current best practice versus the Gemfile hell that we
have had to maintain up until now.

- Update Travis CI testing configuration from sous-chefs templates
  - Use Chef Delivery Local mode instead of Rake for testing
  - Remove unnecessary files for Rubocop and Foodcritic
  - Get rid of that terrible Gemfile (yay)
  - Use Dokken for testing inside Docker containers
  - Let `cookstyle -a` do it's thing
  - Functional tests for HDFS/YARN on
    - CentOS 7
    - Ubuntu 14.04
    - Ubuntu 16.04
- Updated ChefSpec platform versions to non-deprecated platforms
  - CentOS - 6.9
  - Ubuntu - 14.04
- Updated ChefSpec tests to have better coverage
  - Use `do_nothing` matchers
- Added a hadoop_test cookbook under test/fixtures/cookbooks
  - HDFS and YARN service setup and start
  - Force `systemctl daemon-reload` on systemd-based systems

The hadoop_test cookbook can be updated to add more test coverage.
This is a first pass and I didn't want to take all the fun. ;-)

Signed-off-by: Chris Gianelloni <wolf31o2@gmail.com>

### CODE REVIEW

File: spec/unit/recipes/hadoop_yarn_resourcemanager_spec.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      expect(chef_run.service(pkg)).to do_nothing
+    end
+
+    %w(
+      hdfs-tmpdir
+      hdp22-mapreduce-tarball
+      yarn-app-mapreduce-am-staging-dir
+      yarn-remote-app-log-dir
+    ).each do |exec|
+      it "creates #{exec} resource, but does not run it" do
```
Comment (dereklwood) [inline spec/unit/recipes/hadoop_yarn_resourcemanager_spec.rb]: nitpick: `creates #{exec} execute resource`...

File: .delivery/project.toml [hunk 2]
```diff
@@ -0,0 +1 @@
+remote_file = "https://raw.githubusercontent.com/chef-cookbooks/community_cookbook_tools/master/delivery/project.toml"
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply .delivery/project.toml]: With community cookbooks, we want to continue to follow the community standards, even if they change. This was taken from the `sous-chefs` way of doing it. If you would prefer use a local copy, we can do that. For now, I'm going to leave this using the remote file.

File: .kitchen.dokken.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  name: dokken
+  # These three lines enforce Chef 13 compatibility
+  client_rb:
+    treat_deprecation_warnings_as_errors: true
+    resource_cloning: false
+
+platforms:
+  - name: centos-6
+    driver:
+      chef_version: '12.21.3'
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply .kitchen.dokken.yml]: Probably. Unfortunately, Chef's site isn't being very helpful. I think the issue was something to do with deprecation on these older platforms under newer Chef.

File: .kitchen.dokken.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    attributes:
+      hadoop:
+        hdfs_site:
+         'dfs.datanode.max.transfer.threads': 4096
+      hive:
+        hive_site:
+         'hive.support.concurrency': true
+         'hive.zookeeper.quorum': 'localhost'
+      java:
+        jdk_version: 7
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply .kitchen.dokken.yml]: I'll do that separately.

**Pull request merged**